src/wallet/rpcwallet.cpp

```diff
@@ -3463,57 +3476,112 @@ UniValue z_listreceivedbyaddress(const UniValue& params, bool fHelp)
     }
 
     UniValue result(UniValue::VARR);
-    std::vector<SproutNoteEntry> sproutEntries;
-    std::vector<SaplingNoteEntry> saplingEntries;
-    pwalletMain->GetFilteredNotes(sproutEntries, saplingEntries, fromaddress, nMinDepth, false, false);
 
     std::set<std::pair<PaymentAddress, uint256>> nullifierSet;
     auto hasSpendingKey = boost::apply_visitor(HaveSpendingKeyForPaymentAddress(pwalletMain), zaddr);
     if (hasSpendingKey) {
         nullifierSet = pwalletMain->GetNullifiersForAddresses({zaddr});
     }
 
+    NotesFilters filters;
+    filters.minDepth = nMinDepth;
+    filters.ignoreSpent = false;
+    filters.requireSpendingKey = false;
+    filters.cwallet = pwalletMain;
+
     if (boost::get<libzcash::SproutPaymentAddress>(&zaddr) != nullptr) {
-        for (SproutNoteEntry & entry : sproutEntries) {
+        const auto pa = *boost::get<libzcash::SproutPaymentAddress>(&zaddr);
+        const auto itr_sprout = pwalletMain->GetNotesByType(pa.GetHash(), NoteType::sprout, timestamp);
+        auto itr = itr_sprout.begin();
+
+        while (itr != itr_sprout.end() && result.size() < limit) {
+            auto wtx = pwalletMain->mapWallet[itr->hash];
+            const JSOutPoint jsop = *itr->jsop;
+            const uint64_t ts = itr->timestamp;
+            const uint256 txid = itr->hash;
+
+            filters.wtx = &wtx;
+            filters.nd_sprout = pwalletMain->mapWallet[txid].mapSproutNoteData[jsop];
+            filters.jsop = jsop;
+            filters.pa_sprout = pa;
```

Comment:
We tend to use abbreviations more sparingly than this in field names (e.g. `sproutNoteData`, `jsOutPoint`, `sproutAddress`). But this may be academic given the refactoring I suggest below.

Reply:
https://github.com/zcash/zcash/pull/4399/commits/1431921924f32190b4f34f7c8cf1ea81d2957549